docs/whats-new/preview-4.md

```diff
@@ -0,0 +1,644 @@
+        "redis": "local-connection-string"
+    }
+}
+```
+
+Manifest representation:
+
+```json
+{
+  "resources": {
+    "redis": {
+      "type": "parameter.v0",
+      "connectionString": "{redis.value}",
+      "value": "{redis.inputs.value}",
+      "inputs": {
+        "value": {
+          "type": "string",
+          "secret": true
+        }
+      }
+    }
+  }
+}
+```
+
+## New idioms
+
+The following section outlines several new idioms that have been introduced in .NET Aspire preview 4. These idioms are designed to make it easier to model common scenarios when building the application model.
+
+### The `DistributedApplicationExecutionContext`
+
+The `DistributedApplicationExecutionContext` is a new type that provides information about the current execution context. It can be used to determine if the application is being orchestrated running locally or if it is being use to published the manifest.
+
+This can be useful when building the application model. For example, you might want to use a different message broker when running locally than when deploying.
+
+```csharp
+var builder = DistributedApplication.CreateBuilder(args);
+
+if (builder.ExecutionContext.IsPublishMode)
+{
+    // Do something when running in publish mode
+}
+else
+{
+    // Do something when running locally
+}
+```
+
+The `DistributedApplicationExecutionContext` is also available in the DI container and can be used to determine the execution context when resolving services.
+
+### `PublishAs`, `RunAs` and `As`
+
+.NET Aspire preview 4 introduces new idioms for describing common scenarios for modeling how resources are used when running locally and when deploying. While it's possible to model these scenarios using the `DistributedApplicationExecutionContext`, the new idioms make it easier to express these common scenarios.
+
+- `RunAsXX`: Only affects the model when running locally.
+- `PublishAsXX`: Only affects the model when publishing the manifest.
+- `AsXX`: Affects the model both when running locally and when publishing the manifest.
+
+#### Examples
+
+The following logic will use a redis container locally and prompt for
+the connection string when deploying.
+
+```csharp
+var builder = DistributedApplication.CreateBuilder(args);
+
+var redis = builder.AddRedis("redis")
+                   .PublishAsConnectionString();
+
+builder.AddProject<Projects.WebApplication1>("api")
+       .WithReference(redis);
+
+builder.Build().Run();
+```
+
+## General application model improvements
+
+### Changing container properties
+
+We added some new methods to tweak container images, tags and volumes.
+
+Here's an example of using the redis image from Microsoft's container registry instead of the default image from DockerHub.
+
+```csharp
+var builder = DistributedApplication.CreateBuilder(args);
+
+var redis = builder.AddRedis("redis")
+                   .WithImage("mcr.microsoft.com/cbl-mariner/base/redis")
+                   .WithImageTag("6.2-cm2.0");
```

Comment:
@mitchdenny looks like this doesn't work https://github.com/dotnet/aspire/issues/2580